flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -676,11 +682,15 @@ private List<StreamEdge> createChain(
                 config.setChainIndex(chainIndex);
                 config.setOperatorName(streamGraph.getStreamNode(currentNodeId).getOperatorName());
 
+                LinkedHashSet<NonChainedOutput> transitiveOutputs = new LinkedHashSet<>();
                 for (StreamEdge edge : transitiveOutEdges) {
-                    connect(startNodeId, edge);
+                    NonChainedOutput output =
+                            opIntermediateOutputs.get(edge.getSourceId()).get(edge);
+                    transitiveOutputs.add(output);
+                    connect(startNodeId, edge, output);
                 }
 
-                config.setOutEdgesInOrder(transitiveOutEdges);
+                config.setStreamOutputsInOrder(new ArrayList<>(transitiveOutputs));
```

Comment:
Should here also need the de-duplicated outputs?

It looks to me this part should be the list of outputs for the whole vertex, and the non-chained output lists is the per-operator one. Currently it seems at the TM side we first create all the outputs without deduplication, then we connect the operators deduplicated outputs to the chosen ones. This should work, but there seems a bit inconsistent to me

Reply:
The outputs are already de-duplicated. Because ```transitiveOutputs``` is a set and different edges share the same output use the same ```NonChainedOutput``` object. To make is more clear. I will add ```equal``` and ```hashCode``` to the ```NonChainedOutput``` class.